### PR TITLE
feat(manifest): more robust status handling

### PR DESCRIPTION
**Pull request by lwander**

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+  @Autowired
+  OortService oortService;
+
+  @Autowired
+  ObjectMapper objectMapper;
+
+  @Override
+  public long getBackoffPeriod() {
+    return TimeUnit.SECONDS.toMillis(1);
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: This seems a little overly aggressive ... for most of the AWS-operations we've backed this off to 5-10s.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  public TaskResult execute(Stage stage) {
+    String account = getCredentials(stage);
+    String name = (String) stage.getContext().get("manifest.name");
+    String location = (String) stage.getContext().get("manifest.location");
+    String id = String.format("'%s' in '%s' for account %s", name, location, account);
+    List<String> messages = new ArrayList<>();
+
+    Response response;
+    try {
+      response = oortService.getManifest(account, location, name);
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Is there a reason you're returning `Response` from `getManifest` vs. `Manifest`? 

I can't see a need for Response unless you were wanting to stream a response or inspect headers. There's a lot of legacy using `Response` to no great benefit IMO.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    } catch (RetrofitError e) {
+      log.warn("Unable to read manifest {}", id, e);
+      return new TaskResult(ExecutionStatus.RUNNING, new HashMap<>(), new HashMap<>());
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: The `new HashMap<>()` should be redundant, there's a status-only constructor.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    if (response.getStatus() != 200) {
+      return new TaskResult(ExecutionStatus.RUNNING, new HashMap<>(), new HashMap<>());
+    }
+
+    Manifest manifest;
+    try {
+      manifest = objectMapper.readValue(response.getBody().in(), Manifest.class);
+    } catch (IOException e) {
+      log.error("Failed to read " + id, e);
+      throw new RuntimeException(e);
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Why not throw exception w/ message and cause? At this point, you're likely to get the stack logged twice ... one for the explicit log here, and another when the upstream catches the `RuntimeException` and logs.

You may also want to consider including the `executionId` in the error message, we've found it difficult to debug issues in general without it.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    Manifest.Status status = manifest.getStatus();
+    if (status.getStable() == null || !status.getStable().isState()) {
+      messages.add(status.getStable().getMessage());
+      Map<String, Object> context = new ImmutableMap.Builder<String, Object>()
+          .put("stableMessages", messages)
+          .build();
+
+      return new TaskResult(ExecutionStatus.RUNNING, context, new HashMap<>());
```
Comment (ajordens) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: This is going to overwrite any `stableMessages` that were added previously, is that likely to be an issue in reality?

Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: That won't be a problem - however - 

I thought that with the new non-global context that @robfletcher has been working on migrating to we could look up any stage's outputs, and even aggregate across all stages that emit a certain key?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java (hunk 4 above)
Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Sounds good

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java (hunk 3 above)
Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Ah neat

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java (hunk 2 above)
Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Didn't know that was option, `Manifest` sound cleaner.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java (hunk 1 above)
Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Ok - will do

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java (hunk 5 above)
Comment (danielpeach) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: I think he might be asking about the sequence of `stableMessages` generated by retries of this task. Which might be interesting.

Comment (lwander) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/manifest/UpdateManifestWaitForStableTask.java]: Ahh, that's definitely worth overwriting

**Pull request merged**